File: MusicFestival/Repository/ArtistDBRepository.cs
﻿using log4net;
using MPP_MusicFestival.Domain;
using MPP_MusicFestival.Repository;
using System;
using System.Collections.Generic;
using System.Data;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace MusicFestival.Repository
{
    class ArtistDBRepository : ArtistRepository
    {

        private static readonly ILog log = LogManager.GetLogger("ArtistDBRepository");

        IDictionary<String, string> props;

        public ArtistDBRepository(IDictionary<String, string> props)
        {
            log.Info("Creating ArtistDBRepository ");
            this.props = props;
        }

        public void delete(long id)
        {
            IDbConnection con = DBUtils.getConnection(props);
            using (var comm = con.CreateCommand())
            {
                comm.CommandText = "delete from Artist where id=@id";
                IDbDataParameter paramId = comm.CreateParameter();
                paramId.ParameterName = "@id";
                paramId.Value = id;
                comm.Parameters.Add(paramId);
                var dataR = comm.ExecuteNonQuery();
                //if (dataR == 0)
                  //  throw new RepositoryException("No task deleted!");
            }
        }
    

        public IEnumerable<Artist> findAll()
        {
            IDbConnection con = DBUtils.getConnection(props);
            IList<Artist> artists = new List<Artist>();
            using (var comm = con.CreateCommand())
            {
                comm.CommandText = "select * from Artist";

                using (var dataR = comm.ExecuteReader())
                {
                    while (dataR.Read())
                    {
                        long idA = dataR.GetInt32(0);
                        String first_name = dataR.GetString(1);
                        String last_name = dataR.GetString(2);
                        int age = dataR.GetInt32(3);
                        String origin_country = dataR.GetString(4);
                        Artist artist = new Artist(idA, first_name, last_name, age, origin_country);
                       
                        artists.Add(artist);
                    }
                }
            }

            return artists;
        }

        public Artist findOne(long id)
        {
            log.InfoFormat("Entering findOne with value {0}", id);
            IDbConnection con = DBUtils.getConnection(props);

            using (var comm = con.CreateCommand())
            {
                comm.CommandText = "select * from Artist where id=@id";
                IDbDataParameter paramId = comm.CreateParameter();
                paramId.ParameterName = "@id";
                paramId.Value = id;
                comm.Parameters.Add(paramId);

                using (var dataR = comm.ExecuteReader())
                {
                    if (dataR.Read())
                    {
                        long idA = dataR.GetInt32(0);
                        String first_name = dataR.GetString(1);
                        String last_name = dataR.GetString(2);
                        int age = dataR.GetInt32(3);
                        String origin_country = dataR.GetString(4);
                        Artist artist = new Artist(idA, first_name, last_name, age, origin_country);

                        log.InfoFormat("Exiting findOne with value {0}", artist);
                        return artist;
                    }
                }
            }
            log.InfoFormat("Exiting findOne with value {0}", null);
            return null;
        }

        public void save(Artist entity)
        {
            var con = DBUtils.getConnection(props);

            using (var comm = con.CreateCommand())
            {
                comm.CommandText = "insert into Artist(first_name, last_name, age, origin_country) values (@first_name, @last_name, @age, @origin_country)";
                var paramfName = comm.CreateParameter();
                paramfName.ParameterName = "@first_name";
                paramfName.Value = entity.firstName;
                comm.Parameters.Add(paramfName);

                var paramlName = comm.CreateParameter();
                paramlName.ParameterName = "@last_name";
                paramlName.Value = entity.lastName;
                comm.Parameters.Add(paramlName);

                var paramAge = comm.CreateParameter();
                paramAge.ParameterName = "@age";
                paramAge.Value = entity.age;
                comm.Parameters.Add(paramAge);

                var paramOrigin = comm.CreateParameter();
                paramOrigin.ParameterName = "@origin_country";
                paramOrigin.Value = entity.originCountry;
                comm.Parameters.Add(paramOrigin);

                var result = comm.ExecuteNonQuery();
                //if (result == 0)
                  //  throw new RepositoryException("No task added !");
            }
        }

        public void update(long id, Artist entity)
        {
            throw new NotImplementedException();
        }
    }
}

File: MPP_MusicFestival_Java/src/main/java/repository/ArtistDBRepository.java
package repository;

import domain.Artist;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.ArrayList;
import java.util.List;
import java.util.Properties;

public class ArtistDBRepository implements ArtistRepository {

    private JdbcUtils dbUtils;
    private static final Logger logger = LogManager.getLogger();

    public ArtistDBRepository(Properties props) {
        logger.info("Initializing ArtistDBRepository with properties: {} ", props);
        dbUtils = new JdbcUtils(props);
    }


    @Override
    public void save(Artist entity) {
        logger.traceEntry("saving artist {} ", entity);
        Connection con = dbUtils.getConnection();

        try (PreparedStatement preStmt = con.prepareStatement("insert into Artist(first_name, last_name, age, origin_country) values (?,?,?,?)")) {
            preStmt.setString(1, entity.getFirstName());
            preStmt.setString(2, entity.getLastName());
            preStmt.setInt(3, entity.getAge());
            preStmt.setString(4, entity.getOriginCountry());
            preStmt.executeUpdate();
        } catch (SQLException ex) {
            logger.error(ex);
            System.out.println("Error DB " + ex);
        }
        logger.traceExit();
    }

    @Override
    public void delete(Long aLong) {
        logger.traceEntry("deleting artist with {}", aLong);
        Connection con = dbUtils.getConnection();

        try (PreparedStatement preStmt = con.prepareStatement("delete from Artist where id=?")) {
            preStmt.setLong(1, aLong);
            preStmt.executeUpdate();
        } catch (SQLException ex) {
            logger.error(ex);
            System.out.println("Error DB " + ex);
        }
        logger.traceExit();
    }

    @Override
    public void update(Long aLong, Artist entity) {
    }

    @Override
    public Artist findOne(Long aLong) {
        logger.traceEntry("finding artist with id {} ", aLong);
        Connection con = dbUtils.getConnection();

        try (PreparedStatement preStmt = con.prepareStatement("select * from Artist where id=?")) {
            preStmt.setLong(1, aLong);
            try (ResultSet result = preStmt.executeQuery()) {
                if (result.next()) {
                    long id = result.getLong("id");
                    String first_name = result.getString("first_name");
                    String last_name = result.getString("last_name");
                    int age = result.getInt("age");
                    String origin_country = result.getString("origin_country");

                    Artist artist = new Artist(id, first_name, last_name, age, origin_country);
                    logger.traceExit(artist);
                    return artist;
                }
            }
        } catch (SQLException ex) {
            logger.error(ex);
            System.out.println("Error DB " + ex);
        }
        logger.traceExit("No artist found with id {}", aLong);
        return null;
    }

    @Override
    public List<Artist> findAll() {
        logger.traceEntry();
        Connection con = dbUtils.getConnection();
        List<Artist> artists = new ArrayList<>();
        try (PreparedStatement preStmt = con.prepareStatement("select * from Artist")) {
            try (ResultSet result = preStmt.executeQuery()) {
                while (result.next()) {
                    long id = result.getLong("id");
                    String first_name = result.getString("first_name");
                    String last_name = result.getString("last_name");
                    int age = result.getInt("age");
                    String origin_country = result.getString("origin_country");

                    Artist artist = new Artist(id, first_name, last_name, age, origin_country);
                    artists.add(artist);
                }
            }

        } catch (SQLException ex) {
            logger.error(ex);
            System.err.println("Error DB" + ex);
        }
        logger.traceExit();
        return artists;
    }
}
